crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3723,43 +3759,105 @@ impl StringPartsCollector {
+    }
+    .or_else(|| {
+        // When no appropriate method returns any value other than NotImplemented,
+        // the `==` and `!=` operators will fall back to `is` and `is not`, respectively.
+        // refer to `<https://docs.python.org/3/reference/datamodel.html#object.__eq__>`
+        if matches!(op, RichCompareOperator::Eq | RichCompareOperator::Ne) {
+            Some(KnownClass::Bool.to_instance(db))
+        } else {
+            None
+        }
+    })
+    .ok_or_else(|| CompareUnsupportedError {
         op: op.into(),
-        left_ty: Type::Instance(left),
-        right_ty: Type::Instance(right),
+        left_ty: Type::Instance(left_class),
+        right_ty: Type::Instance(right_class),
     })
 }
 
+/// Performs a membership test (`in` and `not in`) between two instances and returns the resulting type.
+/// The behavior can be customized in Python by implementing `__contains__`, `__iter__`, or `__getitem__` methods.
+/// See `<https://docs.python.org/3/reference/datamodel.html#object.__contains__>`
+/// and `<https://docs.python.org/3/reference/expressions.html#membership-test-details>`
+fn perform_membership_test_comparison<'db>(
+    db: &'db dyn Db,
+    left_class: ClassType<'db>,
+    right_class: ClassType<'db>,
+    op: MembershipTestCompareOperator,
+) -> Result<Type<'db>, CompareUnsupportedError<'db>> {
+    let (left_instance, right_instance) = (Type::Instance(left_class), Type::Instance(right_class));
+
+    let contains_dunder = right_class.class_member(db, "__contains__");
+
+    let compare_result_opt = if contains_dunder.is_unbound() {
+        // If `__contains__` is unbound, the code falls back to checking the `__iter__` method for an iteration-based membership test.
+        right_class
+            .class_member(db, "__iter__")
+            .call(db, &[right_instance])
+            .return_ty(db)
+            .map(|_| KnownClass::Bool.to_instance(db))
+            .or_else(|| {
+                // If `__iter__` is not defined, the code falls back to the old-style iteration protocol,
+                // where Python iterates over natural numbers (0..) until an IndexError is raised or the item is found.
+                right_class
+                    .class_member(db, "__getitem__")
+                    .call(db, &[right_instance, KnownClass::Int.to_instance(db)]) // Simulate item access with an integer index.
+                    .return_ty(db)
+                    .map(|_item_ty| KnownClass::Bool.to_instance(db))
+            })
+    } else {
+        // If `__contains__` is available, it is used directly for the membership test.
+        contains_dunder
+            .call(db, &[right_instance, left_instance])
+            .return_ty(db)
+    };
+
+    compare_result_opt
+        .map(|ty| match op {
+            MembershipTestCompareOperator::In => ty.bool(db).into_type(db),
+            MembershipTestCompareOperator::NotIn => ty.bool(db).negate().into_type(db),
+        })
```

Comment:
I think we need some tests with `__contains__` methods returning e.g. `Literal[True]` or `Literal[False]` to show this logic is correct?

Reply:
I missed commenting earlier! sorry 😢

I think the [new test here](https://github.com/astral-sh/ruff/pull/13903/files#diff-93415f322d4e43d79a593ad9befefe07cf5c53ee929363403cb17f740f5df341R73-R95) should cover the scenario you mentioned!